packages/analytics/amplify_analytics_pinpoint/lib/src/flutter_device_context_info_provider.dart

```diff
@@ -0,0 +1,71 @@
+import 'dart:io' show Platform;
+
+import 'package:amplify_analytics_pinpoint/src/app_info_provider.dart';
+import 'package:amplify_analytics_pinpoint/src/device_info_provider.dart';
+import 'package:amplify_analytics_pinpoint_dart/amplify_analytics_pinpoint_dart.dart';
+import 'package:flutter/material.dart';
+
+class FlutterDeviceContextInfoProvider implements DeviceContextInfoProvider {
+  @override
+  String? get countryCode => WidgetsBinding.instance.window.locale.countryCode;
+
+  @override
+  String? get locale => WidgetsBinding.instance.window.locale.toString();
+
+  @override
+  String? get timezone => DateTime.now().timeZoneName;
+
+  // App Info
+  @override
+  String? get appName => _appName;
+
+  @override
+  String? get appPackageName => _appPackageName;
+
+  @override
+  String? get appVersion => _appVersion;
+
+  // Device Info
+  @override
+  String? get make => _make;
+
+  @override
+  String? get model => _model;
+
+  @override
+  String? get modelVersion => _modelVersion;
+
+  @override
+  String? get platform => _platform;
+
+  @override
+  String? get platformVersion => _platformVersion;
+
+  late final String? _appName;
+  late final String? _appPackageName;
+  late final String? _appVersion;
+
+  late final String? _make;
+  late final String? _model;
+  late final String? _modelVersion;
+  late final String? _platform;
+  late final String? _platformVersion;
+
+  FlutterDeviceContextInfoProvider();
+
+  @override
+  Future<void> getDeviceInfoDetails() async {
```

Comment:
This pattern where all data is null until `getDeviceInfoDetails()` is called seems error prone to me. The consumer of this class has to know to call this before they try to access properties. Seems like an easy thing to mess up when making changes inside the flutter package in the future.

I would suggest something like:

```dart
abstract class DeviceContextInfoProvider {
  Future<DeviceContextInfo> getDeviceInfoDetails();
}

class DeviceContextInfo {
  const DeviceContextInfo({
    required this.countryCode,
    required this.locale,
    required this.timezone,
    required this.appName,
    required this.appPackageName,
    required this.appVersion,
    required this.make,
    required this.model,
    required this.modelVersion,
    required this.platform,
  });

  final String countryCode;
  final String locale;
  final String timezone;
  final String appName;
  final String appPackageName;
  final String appVersion;
  final String make;
  final String model;
  final String modelVersion;
  final String platform;
}

class FlutterDeviceContextInfoProvider implements DeviceContextInfoProvider {

  const FlutterDeviceContextInfoProvider();

  @override
  Future<DeviceContextInfo> getDeviceInfoDetails() async {
    var appInfo = await AppInfoProvider.getAppData();
    var deviceInfo = await DeviceInfoProvider.getDeviceData();
    return DeviceContextInfo(
      countryCode: WidgetsBinding.instance.window.locale.countryCode,
      locale: WidgetsBinding.instance.window.locale.toString(),
      timezone: DateTime.now().timeZoneName,
      appName: appInfo.appName,
      appPackageName: appInfo.packageName,
      appVersion: appInfo.version,
      make: deviceInfo.make,
      model: deviceInfo.model,
      modelVersion: deviceInfo.modelVersion,
      platform: deviceInfo.platform,
    );
  }
}
```

Reply:
Thanks added the changes